### PR TITLE
layer: implement BatchNorm layer

### PR DESCRIPTION
**Pull request by iblis17**
Please review it carefully, no test cases :p 

I just plug this layer to the MNIST example, and set the learning rate of SGD to a crazy number: `10`.
It performs quite fine and better than the original model.

---
Ref: [Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](https://arxiv.org/pdf/1502.03167.pdf)

### CODE REVIEW

**Comment by jekbradbury** [comment]
probably shouldn't call it Normalize, since layer normalization and weight normalization are almost as common as batch normalization these days?

**Comment by MikeInnes** [comment]
Yeah, "BatchNorm" seems to be the, uh, norm for this layer.

I find this docstring a bit confusing in terms of what the example is trying to tell me. I guess the message is that this is normally used before non-linearities other than softmax, but it'd be good to just say that if so.

**Comment by iblis17** [comment]
🆗 ,I rename it to `BatchNorm`.

> I find this docstring a bit confusing in terms of what the example is trying to tell me. I guess the message is that this is normally used before non-linearities other than softmax, but it'd be good to just say that if so.

In the section 3.2 of paper, it put the BatchNorm before the activation function. And for the output softmax layer ... I think we can put it, also. I changed the docstring.

File: src/layers/basic.jl [hunk 1]
```diff
@@ -66,3 +66,44 @@ function Base.show(io::IO, l::Dense)
...
+function (N::BatchNorm)(x)
+  μ  = sum(x, 2) ./ size(x, 2)
+  σ² = sum((x .- μ).^2, 2) ./ size(x, 2)
+  N.γ .* ((x .- μ) ./ sqrt.(σ² .+ 1e-8)) .+ N.β
+end
+
+treelike(BatchNorm)
+
+Base.show(io::IO, l::BatchNorm) =
+  print(io, "Normalize(γ = ", l.γ.data[], ", β = ", l.β.data[], ")")
```
Comment (CarloLucibello) [inline src/layers/basic.jl]: BatchNorm here

**Comment by CarloLucibello** [comment]
I wouldn't know what the best interface for this in Flux would be, but the BatchNorm should be keeping a running mean and variance to be updated in train phase and to be used in test phase. Here a Knet example https://github.com/denizyuret/Knet.jl/pull/140/files#diff-8f43844167d26280cce2ebdd6170d842R39 I wrote some time ago.

In PyTorch you can explicitely differentiate train phase and test phase with model.train() and model.eval()

**Comment by iblis17** [comment]
> I wouldn't know what the best interface for this in Flux would be, but the BatchNorm should be keeping a running mean and variance to be updated in train phase and to be used in test phase.

That's why I use `treelike` in this PR
```
treelike(BatchNorm)
```

and in the model
```julia
m = Chain(..., BatchNorm(), ...)
SGD(params(m))  # optimizer will update it
```

**Comment by iblis17** [comment]
hmm, I read the Algo2 in paper again, I think I misunderstood the dimension of gamma and beta.

**Comment by CarloLucibello** [comment]
I was talking about `mu` and `sigma^2`, you should accumulate them somewhere

**Comment by iblis17** [comment]
@CarloLucibello oh, I got your point, but do we have batching functionality in Flux?
(I do not do batching in the MNIST example

File: src/layers/basic.jl [hunk 2]
```diff
@@ -66,3 +66,44 @@ function Base.show(io::IO, l::Dense)
...
+
+```julia
+julia> m = Chain(
+  Dense(28^2, 32, identity),
+  BatchNorm(),
+  relu,
+  Dense(32, 10, identity),
+  BatchNorm(),
+  softmax)
+Chain(Dense(784, 32), Normalize(γ = 1.0, β = 0.0), NNlib.relu, Dense(32, 10), Normalize(γ = 1.0, β = 0.0), NNlib.softmax)
```
Comment (CarloLucibello) [inline src/layers/basic.jl]: other Normalize left here

**Comment by CarloLucibello** [comment]
I normally handcode batching, as in Knet examples, or there is some batching functionality in some of the JuliaML packages

**Comment by CarloLucibello** [comment]
In any case, in test phase you want to be able to make a prediction also on a single image, and I think this won't give you a good result with the code as it is (the variance becomes zero).

**Comment by iblis17** [comment]
I will wait for the design of training/testing phase being introducing to Flux

**Comment by MikeInnes** [comment]
I just merged #92 which should have everything you need for the train/test distinction.

**Comment by iblis17** [comment]
@CarloLucibello please review the algo :)
and I found that BachNorm layer ate up my memory quickly with the MNIST example. :/

---
TODO
- [ ] add test cases

**Comment by CarloLucibello** [comment]
Two things are missing here:

- the BatchNorm layer should work also in conjunction with convolutional layers. 
- you should keep an array *of fixed size* of moments, one for each batch in the training set, or you can keep just one set of moments and update it with a decaying average. I think we should go with the latter way, which is easier to implement and less memory consuming. 

You can look for inspiration to my PR here https://github.com/denizyuret/Knet.jl/pull/140 (the decaying update is also there but commented out).

**Comment by iblis17** [comment]
More question about moments: this function looks weird 
```julia
# function Base.push!(b::BatchMoments, μ, σ)
#     b.μ = b.momentum .* b.μ .+ (1 - b.momentum) .* b.μ
#     b.σ = b.momentum .* b.σ .+ (1 - b.momentum) .* b.σ
# end
```
Should it be
```julia
b.μ = b.momentum .* μ .+ (1 - b.momentum) .* b.μ
b.σ = b.momentum .* σ .+ (1 - b.momentum) .* b.σ

```
?

**Comment by CarloLucibello** [comment]
Ah, right, stupid me

Il ven 27 ott 2017, 12:36 Iblis Lin <notifications@github.com> ha scritto:

**Comment by CarloLucibello** [comment]
A good default value for the momentum of the running average could be `momentum=0.1`
http://pytorch.org/docs/master/nn.html

**Comment by iblis17** [comment]
We don't have convolution layer yet. Maybe add it later and in another PR ?

File: src/layers/normalisation.jl [hunk 3]
```diff
@@ -43,3 +43,72 @@ function (a::Dropout)(x)
...
+    σ = BN.σ
+  else
+    m = size(x, 2)  # batch size
+    μ = sum(x, 2) ./ m
+    σ = sqrt.(sum((x .- μ).^2, 2) ./ (m - 1) .+ BN.ϵ)
+
+    # update moving mean/std
+    mtm = BN.momentum
+    BN.μ = mtm .* μ.data .+ (1 - mtm) .* BN.μ
+    BN.σ = mtm .* σ.data .+ (1 - mtm) .* BN.σ
```
Comment (CarloLucibello) [inline src/layers/normalisation.jl]: momentum and epsilong being Float64 makes the output Array{Float64} inven if the input is Array{Float32}. A
solution could be to define 
```julia
T = eltype(x)
 ϵ = T(BN.ϵ)
mtm = T(BN.momentum)

```

File: src/layers/normalisation.jl [hunk 4]
```diff
@@ -43,3 +43,72 @@ function (a::Dropout)(x)
...
+```
+"""
+mutable struct BatchNorm{F,V}
+  λ::F  # activation function
+  β::V  # bias
+  γ::V  # scale
+  μ     # moving mean
+  σ     # moving std
+  ϵ::Float64
+  momentum::Float64
```
Comment (CarloLucibello) [inline src/layers/normalisation.jl]: don't have to be neccesarily Float64 here, we could parametrix or leave untyped

**Comment by iblis17** [comment]
test case added

**Comment by iblis17** [comment]
currently, it doesn't work with CuArrays.
```julia
ERROR: LoadError: don't know how to handle argument of type Flux.##83#88{Flux.BatchNorm{NNlib.#relu,TrackedArray{…,CuArray{Float32,1}},Float64}}
Stacktrace:
 [1] cudaconvert(::Flux.##83#88{Flux.BatchNorm{NNlib.#relu,TrackedArray{…,CuArray{Float32,1}},Float64}}) at /home/iblis/.julia/v0.6/CUDAnative/src/execution.jl:20
 [2] map at ./tuple.jl:161 [inlined] (repeats 2 times)
 [3] broadcast at ./broadcast.jl:17 [inlined]
 [4] _broadcast! at /home/iblis/.julia/v0.6/CuArrays/src/broadcast.jl:22 [inlined]
 [5] broadcast_t at /home/iblis/.julia/v0.6/CuArrays/src/broadcast.jl:37 [inlined]
 [6] broadcast_c at /home/iblis/.julia/v0.6/CuArrays/src/broadcast.jl:58 [inlined]
 [7] broadcast at ./broadcast.jl:434 [inlined]
 [8] tracked_broadcast(::Function, ::TrackedArray{…,CuArray{Float32,1}}, ::TrackedArray{…,CuArray{Float32,2}}, ::TrackedArray{…,CuArray{Float32,2}}, ::TrackedArray{…,CuArray{Float32,2}}, ::TrackedArray{…,CuArray{Float32,1}}) at /home/iblis/.julia/v0.6/Flux/src/tracker/lib.jl:100
 [9] (::Flux.BatchNorm{NNlib.#relu,TrackedArray{…,CuArray{Float32,1}},Float64})(::TrackedArray{…,CuArray{Float32,2}}) at /home/iblis/.julia/v0.6/Flux/src/layers/normalisation.jl:105
 [10] mapfoldl_impl(::Base.#identity, ::Flux.##57#58, ::CuArray{Float32,2}, ::Array{Any,1}, ::Int64) at ./reduce.jl:46
 [11] (::Flux.Chain)(::CuArray{Float32,2}) at /home/iblis/.julia/v0.6/Flux/src/layers/basic.jl:30
 [12] loss(::CuArray{Float32,2}, ::Flux.OneHotMatrix{CuArray{Flux.OneHotVector,1}}) at /home/iblis/git/Flux/model-zoo/mnist/mnist.jl:24
 [13] macro expansion at /home/iblis/.julia/v0.6/Flux/src/optimise/train.jl:19 [inlined]
 [14] macro expansion at /home/iblis/.julia/v0.6/Juno/src/progress.jl:128 [inlined]
 [15] #train!#112(::Flux.#throttled#14, ::Function, ::Function, ::Base.Iterators.Take{Base.Iterators.Repeated{Tuple{CuArray{Float32,2},Flux.OneHotMatrix{CuArray{Flux.OneHotVector,1}}}}}, ::Flux.Optimise.##58#62) at /home/iblis/.julia/v0.6/Flux/src/optimise/train.jl:18
 [16] (::Flux.Optimise.#kw##train!)(::Array{Any,1}, ::Flux.Optimise.#train!, ::Function, ::Base.Iterators.Take{Base.Iterators.Repeated{Tuple{CuArray{Float32,2},Flux.OneHotMatrix{CuArray{Flux.OneHotVector,1}}}}}, ::Function) at ./<missing>:0
 [17] include_from_node1(::String) at ./loading.jl:576
 [18] include(::String) at ./sysimg.jl:14
 [19] process_options(::Base.JLOptions) at ./client.jl:305
 [20] _start() at ./client.jl:371
while loading /home/iblis/git/Flux/model-zoo/mnist/mnist.jl, in expression starting on line 38
```

Any idea?

**Comment by MikeInnes** [comment]
That's due to broadcast calls like `BN.a .* xs`. If you unpack first (like `a = BN.a; a .* xs`) it should work.

**Comment by iblis17** [comment]
Ok, it finally works with CuArrays.

File: src/layers/normalisation.jl [hunk 5]
```diff
@@ -43,3 +43,80 @@ function (a::Dropout)(x)
...
+
+    ϵ = T(BN.ϵ)
+    m = size(x, 2)  # batch size
+    μ = mean(x, 2)
+    σ = sqrt.(sum((x .- μ).^2, 2) ./ m .+ ϵ)
+
+    # update moving mean/std
+    mtm = T(BN.momentum)
+    BN.μ = (1 - mtm) .* BN.μ .+ mtm .* μ.data
+    BN.σ = (1 - mtm) .* BN.σ .+ mtm .* σ.data .* m ./ (m - 1)
```
Comment (CarloLucibello) [inline src/layers/normalisation.jl]: I think this should be
BN.μ = mtm .* BN.μ .+ (1-mtm ).* μ.data

Comment (iblis17) [reply src/layers/normalisation.jl]: Given this torch implementation: https://github.com/pytorch/pytorch/blob/502aaf39cf4a878f9e4f849e5f409573aa598aa9/aten/src/THNN/generic/BatchNormalization.c#L46

and I use the `momentum = .1` as default value which is same as pytorch.

So... your version (` BN.μ = mtm .* BN.μ .+ (1-mtm ).* μ.data`) should be paired with `mtm = 0.9` as default value.

Comment (CarloLucibello) [reply src/layers/normalisation.jl]: right, but maybe we should go with the usual meaning for momentum, as used in optimizers, and other BN implementations e.g.
http://gluon.mxnet.io/chapter04_convolutional-neural-networks/cnn-batch-norm-scratch.html
and default to 0.9
Anyway, given there is a precedent in pytorch, any choiche is allright

Comment (iblis17) [reply src/layers/normalisation.jl]: > maybe we should go with the usual meaning for momentum, as used in optimizers, and other BN implementations

So... here we have two options here

1. `mtm * new_vector + ...`
2. `mtm * old + (1-mtm) * new`

In my current code is option (1), right?

Let's dig into Flux's momentum:
https://github.com/FluxML/Flux.jl/blob/dc1f08a709d1dc9e90fba84b0dc446a4c74664de/src/optimise/optimisers.jl#L8-L11

So I think the Flux's optimizer choose option (1).

Next, Let's see the implementation of Gluon. In the doc you show `momentum`'s default value is `0.9` and
```python
    # store the moving statistics by their scope_names, inplace
    if scope_name not in _BN_MOVING_MEANS:
        _BN_MOVING_MEANS[scope_name] = mean
    else:
        _BN_MOVING_MEANS[scope_name] = _BN_MOVING_MEANS[scope_name] * momentum + mean * (1.0 - momentum)
    if scope_name not in _BN_MOVING_VARS:
        _BN_MOVING_VARS[scope_name] = variance
    else:
        _BN_MOVING_VARS[scope_name] = _BN_MOVING_VARS[scope_name] * momentum + variance * (1.0 - momentum)
```
And I think they choose option (2).

I will vote for being consistent with Flux's source code, which is option (1), with default value as `0.1`.
I can add more note in the docsting if you want

**Comment by MikeInnes** [comment]
This is an excellent PR. Thanks a lot @iblis17!

**Pull request merged**